### PR TITLE
SAI proposal for qinq to vxlan

### PR DESCRIPTION
**Pull request by lguohan**

### CODE REVIEW

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+```
+
+### Vxlan to QinQ
+
+
+Create tunnel termination
+
+```
+sai_status_t create_tunnel_termination(
+sai_object_id_t oid,
```
Comment (pavel-shirshov) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: oid of what object?

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+### Vxlan to QinQ
+
+
+Create tunnel termination
+
+```
+sai_status_t create_tunnel_termination(
+sai_object_id_t oid, 
+sai_ip4_t ipv4,
```
Comment (pavel-shirshov) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: Is this tunnel source ip or tunnel destination ip?

**Comment by kcudnik** [comment]
test this please

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+```
+    typedef enum _sai_router_interface_attr_t
+    {
+        /**
+         * @brief Outer Vlan
+         *
+         * @type sai_uint16_t
+         * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+         * @condition SAI_ROUTER_INTERFACE_ATTR_TYPE == SAI_ROUTER_INTERFACE_TYPE_QINQ_PORT;
```
Comment (kcudnik) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: @isvlan tag is needed for uint16 types

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+}
+
+sai_object_id_t create_decap_tunnel_map()
+{
+    sai_status_t status;
+    sai_attribute_t attr;
+    std::vector<sai_attribute_t> tunnel_map_attrs;
+
+    attr.id = SAI_TUNNEL_MAP_ATTR_TYPE;
+    attr.value.s32 = SAI_TUNNEL_MAP_TYPE_VNI_VIRTUAL_ROUTER_ID;
```
Comment (tushar-ty) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: SAI_TUNNEL_MAP_TYPE_VNI_TO_VIRTUAL_ROUTER_ID

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+![qinq-to-vxlan](figures/qinq-to-vxlan.png "Figure 2: QinQ to VxLAN usage scenario")
+__Figure 1: QinQ to VxLAN usage scenario__
+ 
+The packet processing pipeline is below:
+![qinq-to-vxlan-pipeline](figures/qinq-to-vxlan-pipeline.png "Figure 2: QinQ to VxLAN SAI Pipeline")
+__Figure 2: QinQ to VxLAN SAI Pipeline__
+ 
+## Packet Format ##
+
+- QinQ -> vxlan: For the vxlan packet , the inner L2 header is a dummy header without vlan header.
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: yes, we can.

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+
+```
+sai_status_t create_router_interface_qinq(
+    sai_object_id_t overlay_router_id, 
+    uint32_t out_vlan, 
+    uint32_t in_vlan, 
+    sai_object_id_t *router_intf)
+{
+    sai_status_t status;
+    sai_attribute_t intf_attrs[3];
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: fixed

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+    intf_attrs[0].value.oid = overlay_router_id;
+
+    intf_attrs[1].id = SAI_ROUTER_INTERFACE_ATTR_TYPE;
+    intf_attrs[1].value.s32 = SAI_ROUTER_INTERFACE_TYPE_QINQ_PORT;
+
+    intf_attrs[2].id = SAI_ROUTER_INTERFACE_ATTR_OUTER_VLAN_ID;
+    intf_attrs[2].value.u16 = outer_vlan;
+
+    intf_attrs[2].id = SAI_ROUTER_INTERFACE_ATTR_INNER_VLAN_ID;
+    intf_attrs[2].value.u16 = inner_vlan;
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: fixed

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+```
+sai_object_id_t create_encap_tunnel_map()
+{
+    sai_status_t status;
+    sai_attribute_t attr;
+    std::vector<sai_attribute_t> tunnel_map_attrs;
+
+    attr.id = SAI_TUNNEL_MAP_ATTR_TYPE;
+    attr.value.s32 = SAI_TUNNEL_MAP_TYPE_VIRTUAL_ROUTER_ID_TO_VNI;
+    tunnel_attrs.push_back(attr);
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: fixed

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+    sai_object_id_t router_id, 
+    sai_uint32_t vni)
+{
+    sai_status_t status;
+    sai_attribute_t attr;
+    std::vector<sai_attribute_t> tunnel_map_entry_attrs;
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_TUNNEL_MAP_TYPE;
+    attr.value.s32 = SAI_TUNNEL_MAP_TYPE_VIRTUAL_ROUTER_ID_TO_VNI;
+    tunnel_attrs.push_back(attr);
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: fixed

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_VIRTUAL_ROUTER_ID_KEY;
+    attr.value.oid = router_id;
+    tunnel_attrs.push_back(attr);
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_VNI_VALUE;
+    attr.value.oid = vni;
+    tunnel_attrs.push_back(attr);
+
+    sai_tunnel_api->create_tunnel_map(tunnel_map_id, 1, tunnel_map_entry_attrs);
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: fixed

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+
+sai_object_id_t create_decap_tunnel_map()
+{
+    sai_status_t status;
+    sai_attribute_t attr;
+    std::vector<sai_attribute_t> tunnel_map_attrs;
+
+    attr.id = SAI_TUNNEL_MAP_ATTR_TYPE;
+    attr.value.s32 = SAI_TUNNEL_MAP_TYPE_VNI_VIRTUAL_ROUTER_ID;
+    tunnel_attrs.push_back(attr);
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: fixed

File: doc/SAI-Proposal-QinQ-VXLAN.md (hunk 4 above)
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: fixed

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+    sai_object_id_t router_id, 
+    sai_uint32_t vni)
+{
+    sai_status_t status;
+    sai_attribute_t attr;
+    std::vector<sai_attribute_t> tunnel_map_entry_attrs;
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_TUNNEL_MAP_TYPE;
+    attr.value.s32 = SAI_TUNNEL_MAP_TYPE_VNI_TO_VIRTUAL_ROUTER_ID;
+    tunnel_attrs.push_back(attr);
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: fixed

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_VNI_KEY;
+    attr.value.oid = router_id;
+    tunnel_attrs.push_back(attr);
+
+    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_VIRTUAL_ROUTER_VALUE;
+    attr.value.oid = vni;
+    tunnel_attrs.push_back(attr);
+
+    sai_tunnel_api->create_tunnel_map(tunnel_map_id, 1, tunnel_map_entry_attrs);
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: fixed

File: doc/SAI-Proposal-QinQ-VXLAN.md (hunk 1 above)
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: tunnel oid

File: doc/SAI-Proposal-QinQ-VXLAN.md (hunk 2 above)
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: dest ip.

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+        
+        /** TUNNEL Map VNI to Virtual Router ID */
+        SAI_TUNNEL_MAP_TYPE_VNI_TO_VIRTUAL_ROUTER_ID = 0x00000006,
+
+        /** TUNNEL Map Virtual Router ID to VNI */
+        SAI_TUNNEL_MAP_TYPE_VIRTUAL_ROUTER_ID_TO_VNI = 0x00000007,
+    } sai_tunnel_map_type_t;
+```
+
+- New tunnel map entry
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: added

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+
+### QinQ to Vxlan example
+
+```
+# create qinq router interface
+status = create_router_interface_qinq(virtual_router_id, 100, 1, &rif);
+
+# create two tunnels for a same virtual router
+status = create_tunnel(virtual_router_id, 2000, &tunnel_id_1);
+status = create_tunnel(virtual_router_id, 2001, &tunnel_id_2);
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: in case they want to send to different vni

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 16]
```diff
@@ -0,0 +1,429 @@
...
+
+# create two tunnel nexthops on the virtual router
+status = create_nexthop_tunnel(virtual_router_id, "10.10.10.1", tunnel_id_1, &nexthop_id_1);
+status = create_nexthop_tunnel(virtual_router_id, "20.20.20.1", tunnel_id_2, &nexthop_id_2);
```
Comment (lguohan) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: yes, it will work, but this is just another use case.

**Comment by lguohan** [comment]
retest this please

**Pull request merged**